server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -138,33 +167,82 @@ public static ElasticsearchException validate(
                 // checking if flat expression matched remotely
                 for (String remoteExpression : remoteExpressions) {
                     String[] splitResource = splitQualifiedResource(remoteExpression);
-                    ElasticsearchException exception = checkSingleRemoteExpression(
+                    var projectAlias = splitResource[0];
+                    var resource = splitResource[1];
+
+                    ElasticsearchException remoteException = checkSingleRemoteExpression(
                         remoteResolvedExpressions,
-                        splitResource[0], // projectAlias
-                        splitResource[1], // resource
+                        projectAlias,
+                        resource,
                         remoteExpression,
                         indicesOptions
                     );
-                    if (exception == null) {
+                    if (remoteException == null) {
                         // found flat expression somewhere
                         foundFlat = true;
                         break;
                     }
-                    if (false == isUnauthorized && exception instanceof ElasticsearchSecurityException) {
+                    if (false == isUnauthorized && remoteException instanceof ElasticsearchSecurityException) {
                         isUnauthorized = true;
+                        if (remoteAuthorizationExceptions == null) {
+                            remoteAuthorizationExceptions = new HashMap<>();
+                            remoteUnauthorizedIndices = new HashMap<>();
+                        }
+                        remoteAuthorizationExceptions.putIfAbsent(projectAlias, remoteException);
+                        remoteUnauthorizedIndices.compute(projectAlias, (k, v) -> {
+                            if (v == null) {
+                                v = new ArrayList<>();
+                            }
+                            v.add(resource);
+                            return v;
+                        });
                     }
                 }
                 if (foundFlat) {
                     continue;
                 }
                 if (isUnauthorized) {
-                    return localException;
+                    if (localAuthorizationException == null) {
+                        localAuthorizationException = localException;
+                        localUnauthorizedIndices = new ArrayList<>();
+                    }
+                    localUnauthorizedIndices.add(originalExpression);
+                    continue;
                 }
-                return new IndexNotFoundException(originalExpression);
+                notFoundException = new IndexNotFoundException(originalExpression);
```

Comment:
Nit: Can we remove `continue` and place `notFoundException = new IndexNotFoundException(originalExpression)` in an `else` branch which reads better and more consistent with how it is handled in the other branch?